src/global/reco/ParticleFlow_factory.cc

```diff
@@ -0,0 +1,126 @@
+    std::vector<const edm4eic::ClusterCollection*> vecHCalInput(m_const.nCaloPairs);
+
+    // grab and organize input collections
+    TrkInput trkInput;
+    for (const std::string& input_tag : GetInputTags()) {
+
+      // check if input is a track collection
+      if (input_tag.find("ReconstructedChargedParticles") != std::string::npos) {
+        trkInput.first = static_cast<const edm4eic::ReconstructedParticleCollection*>(event->GetCollectionBase(input_tag));
+        m_log->debug("Found input track collection {}", input_tag);
+        continue;
+      }
+
+      // check if input is a track projection collection
+      if (input_tag.find("TrackProjection") != std::string::npos) {
+        trkInput.second = static_cast<const edm4eic::TrackSegmentCollection*>(event->GetCollectionBase(input_tag));
+        m_log->debug("Found input track projection collection {}", input_tag);
+        continue;
+      }
+
+      // check if input is a calo collection
+      bool isECalCollect = (input_tag.find("Ecal") != std::string::npos);
+      bool isHCalCollect = ((input_tag.find("Hcal") != std::string::npos) || (input_tag.find("HCAL") != std::string::npos));
+      bool endsInClust   = (input_tag.find("Clusters") != std::string::npos);
+      if ((isECalCollect || isHCalCollect) && endsInClust) {
+
+        // set location in input vectors
+        size_t iInput = -1;
+        try {
+          iInput = m_mapCaloInputToIndex.at(input_tag);
+        } catch (std::out_of_range &out) {
+          m_log->error("Using unknown calo collection {} as input to particle flow!", input_tag);
+          throw JException(out.what());
+        }
+        m_log->debug("Assigned input calo collection {} an index of {}", input_tag, iInput);
+
+        // grab collections
+        if (isECalCollect) vecECalInput[iInput] = static_cast<const edm4eic::ClusterCollection*>(event->GetCollectionBase(input_tag));
+        if (isHCalCollect) vecHCalInput[iInput] = static_cast<const edm4eic::ClusterCollection*>(event->GetCollectionBase(input_tag));
+      }
+    }  // end input tag loop
+
+    // collect input calo collections into ECal-HCal pairs
+    std::vector<CaloInput> vecCaloInput(m_const.nCaloPairs);
+    for (size_t iCaloPair = 0; iCaloPair < m_const.nCaloPairs; iCaloPair++) {
+      vecCaloInput[iCaloPair] = std::make_pair(vecECalInput[iCaloPair], vecHCalInput[iCaloPair]);
+    }
+
+    // prepare vector of pairs of ECal-HCal IDs
+    VecCaloIDs vecCaloIDs(m_const.nCaloPairs);
+    for (size_t iCaloPair = 0; iCaloPair < m_const.nCaloPairs; iCaloPair++) {
+      const uint32_t idECal = m_geoSvc -> detector() -> constant<uint32_t>(m_mapIndexToECalID[iCaloPair]);
+      const uint32_t idHCal = m_geoSvc -> detector() -> constant<uint32_t>(m_mapIndexToHCalID[iCaloPair]);
+      vecCaloIDs[iCaloPair] = std::make_pair(idECal, idHCal);
+    }
+
+    // run algorithm
+    auto pf_objects = m_pf_algo.process(trkInput, vecCaloInput, vecCaloIDs);
```

Comment:
We are moving to make this all a lot simpler so that what goes into this process is only (and can only be) the collections you specified in the reco.cc generation. Nothing more. If you put all this logic into the factory, someone else is going to have to strip it out again in a few weeks/months.